File: src/Chauffeur.Deliverables.Tests/MockTextWriter.cs
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;

namespace Chauffeur.Deliverables.Tests
{
    class MockTextWriter : TextWriter
    {
        private readonly List<string> messages;
        public MockTextWriter()
        {
            messages = new List<string>();
        }

        public IEnumerable<string> Messages { get { return messages; } }

        public override System.Text.Encoding Encoding { get { return System.Text.Encoding.Default; } }

        public override async Task WriteLineAsync(string value)
        {
            messages.Add(value);
            await Task.FromResult(value);
        }
    }
}

File: src/Chauffeur.Tests/MockTextWriter.fs
namespace Chauffeur.Tests

open System.IO
open System.Threading.Tasks
open System.Diagnostics

type MockTextWriter() =
    inherit TextWriter()

    let mutable messages = []

    member x.Messages = messages

    override x.Flush() =
        messages <- []

    override x.Encoding = System.Text.Encoding.Default

    override x.WriteLineAsync value =
        Debug.WriteLine value
        messages <- value :: messages
        Task.FromResult value :> System.Threading.Tasks.Task
